storage/rocksdb/rdb_locking_iter.cc

```diff
@@ -0,0 +1,278 @@
+    starts from target and extends to some point between target and end_key.
+*/
+void LockingIterator::lock_up_to(bool scan_forward,
+                                 const rocksdb::Slice &target,
+                                 const rocksdb::Slice &end_key) {
+  const int inv = scan_forward ? 1 : -1;
+  auto cmp = m_cfh->GetComparator();
+  bool endp_arg = m_kd->m_is_reverse_cf;
+
+  if (m_have_locked_until &&
+      cmp->Compare(end_key, rocksdb::Slice(m_locked_until)) * inv <= 0) {
+    // We've already locked this range. The following has happened:
+    // - m_iter->key() returned $KEY
+    // - other transaction(s) have inserted row $ROW before the $KEY.
+    // - we got a range lock on [range_start, $KEY]
+    // - we've read $ROW and returned.
+    // Now, we're looking to lock [$ROW, $KEY] but we don't need to,
+    // we already have a lock on this range.
+  } else {
+    if (scan_forward) {
+      m_status = m_txn->GetRangeLock(m_cfh, rocksdb::Endpoint(target, endp_arg),
+                                     rocksdb::Endpoint(end_key, endp_arg));
+    } else {
+      m_status =
+          m_txn->GetRangeLock(m_cfh, rocksdb::Endpoint(end_key, endp_arg),
+                              rocksdb::Endpoint(target, endp_arg));
+    }
+
+    if (!m_status.ok()) return;
+
+    // Save the bound where we locked until:
+    m_have_locked_until = true;
+    m_locked_until.assign(end_key.data(), end_key.size());
+    if (m_lock_count) (*m_lock_count)++;
+  }
+}
+
+/*
+  Lock the range from target till the iterator end point that we are scaning
+  towards. If there's no iterator bound, use index start (or end, depending
+  on the scan direction)
+*/
+void LockingIterator::lock_till_iterator_end(bool scan_forward,
+                                             const rocksdb::Slice &target) {
+  rocksdb::Slice end;
+  uchar buf[Rdb_key_def::INDEX_NUMBER_SIZE];
+  uint size;
+  if (scan_forward) {
+    if (m_read_opts.iterate_upper_bound)
+      end = *m_read_opts.iterate_upper_bound;
+    else {
+      if (m_kd->m_is_reverse_cf)
+        m_kd->get_infimum_key(buf, &size);
+      else
+        m_kd->get_supremum_key(buf, &size);
+
+      DBUG_ASSERT(size == Rdb_key_def::INDEX_NUMBER_SIZE);
+      end = rocksdb::Slice((const char *)buf, size);
+    }
+  } else {
+    if (m_read_opts.iterate_lower_bound)
+      end = *m_read_opts.iterate_lower_bound;
+    else {
+      if (m_kd->m_is_reverse_cf)
+        m_kd->get_supremum_key(buf, &size);
+      else
+        m_kd->get_infimum_key(buf, &size);
+
+      DBUG_ASSERT(size == Rdb_key_def::INDEX_NUMBER_SIZE);
+      end = rocksdb::Slice((const char *)buf, size);
+    }
+  }
+  // This will set m_status accordingly
+  lock_up_to(scan_forward, target, end);
+}
+
+/*
+  Lock the range between [target, (current m_iter position)] and position
+  the iterator on the first record in it.
+
+  @param call_next false means current iterator position is achieved by
```

Comment:
We have `ha_rocksdb::index_next_with_direction_intern` already with `skip_next` that does something similar so I wonder if we should stick to one naming convention.